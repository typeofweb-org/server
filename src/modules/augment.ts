import type { TypeOfWebRequest, TypeOfWebServer, TypeOfWebResponse } from './shared';

export interface TypeOfWebServerMeta {}

export interface TypeOfWebRequestMeta {}

export interface TypeOfWebEvents {
  readonly ':afterResponse': TypeOfWebResponse;
  readonly ':server': TypeOfWebServer;
  readonly ':request': TypeOfWebRequest;
  readonly ':error': unknown;
}
