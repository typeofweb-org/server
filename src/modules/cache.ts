import { invariant } from '../utils/errors';
import { stableJsonStringify } from '../utils/serializeObject';

import type { Json } from '../utils/types';
import type { TypeOfWebCacheConfig } from './shared';
import type CacheManager from 'cache-manager';

const serializeArgs = (args: Json): string => stableJsonStringify(args);

export const createCachedFunction = <Fn extends (...args: readonly Json[]) => any>({
  fn,
  cache,
  cacheInstance,
}: {
  readonly fn: Fn;
  readonly cache: TypeOfWebCacheConfig;
  readonly cacheInstance: CacheManager.Cache;
}): Fn => {
  const ttlMs = 'expireIn' in cache ? cache.expireIn : expireAtToTtlMs(cache.expireAt);

  invariant(ttlMs, 'TTL is undefined - something went wrong');
  invariant(ttlMs > 0, 'TTL<=0 - something went wrong');

  // cache-manager requires ttl in seconds
  const ttl = ttlMs / 1000;

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- ok
  return function (...args) {
    const id = serializeArgs(args);
    return cacheInstance.wrap<ReturnType<Fn>>(
      id,
      () => {
        return fn(...args);
      },
      { ttl },
    );
  } as Fn;
};

function expireAtToTtlMs(expireAt: Exclude<TypeOfWebCacheConfig['expireAt'], undefined>) {
  /* istanbul ignore next */
  const [hours = '00', minutes = '00'] = expireAt.split(':');
  const expireAtDate = new Date();
  expireAtDate.setHours(Number.parseInt(hours));
  expireAtDate.setMinutes(Number.parseInt(minutes));
  const now = new Date();

  if (expireAtDate <= now) {
    // if expire at is in the past then surely it was meant to be the next day
    expireAtDate.setDate(expireAtDate.getDate() + 1);
  }

  return expireAtDate.getTime() - now.getTime();
}
