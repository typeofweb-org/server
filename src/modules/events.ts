import type { EventBus } from './shared';

export const createEventBus = (): EventBus => {
  // @todo
  return {
    emit() {},
    on() {},
    off() {},
  };
};
