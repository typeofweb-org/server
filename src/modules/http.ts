import Bodyparser from 'body-parser';
import Express from 'express';
import Stoppable from 'stoppable';

import type { AppOptions } from './shared';

export const initApp = () => {
  const app = Express();

  app.use(Bodyparser.json());
  app.use(Bodyparser.urlencoded({ extended: true }));
  app.disable('x-powered-by');

  return app;
};

/* istanbul ignore next */
export const listenExpressServer = (app: Express.Application, { port, hostname }: AppOptions) => {
  return new Promise<Stoppable.StoppableServer>((resolve, reject) => {
    app.once('error', reject);

    const server = Stoppable(
      app.listen(port, hostname, () => {
        app.off('error', reject);
        resolve(server);
      }),
      1000,
    );

    const stopServer = (_signal: 'SIGTERM' | 'SIGINT') => {
      server.stop();
    };

    process.on('SIGTERM', () => stopServer('SIGTERM'));
    process.on('SIGINT', () => stopServer('SIGINT'));
  });
};
