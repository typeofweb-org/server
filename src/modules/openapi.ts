import { isSchema, object } from '@typeofweb/schema';
import Prettier from 'prettier';
import { getTypeScriptReader, getOpenApiWriter, makeConverter } from 'typeconv';

import type { TypeOfWebRoute } from './shared';
import type { SomeSchema } from '@typeofweb/schema';
import type { IJsonSchema, OpenAPIV2 } from 'openapi-types';

type RouteSubset = Pick<TypeOfWebRoute, 'method' | 'path' | 'validation'>;

export const routeConfigToOpenApiPathsDefinitions = async (
  route: RouteSubset,
): Promise<Pick<OpenAPIV2.Document, 'definitions' | 'paths'>> => {
  const identifier = routeToIdentifier(route);
  const definitions = (await getOpenApiResultForRoute(identifier, route)).components.schemas;
  const paths = getOpenApiPathForRoute(identifier, route, definitions);

  return {
    definitions,
    paths,
  };
};

export const getOpenApiForRoutes = async (
  routes: readonly RouteSubset[],
  options: { readonly title: string; readonly description?: string; readonly version: string },
): Promise<OpenAPIV2.Document> => {
  const openApiForRoutes = await Promise.all(routes.map(routeConfigToOpenApiPathsDefinitions));
  const { paths, definitions } = openApiForRoutes.reduce((acc, el) => {
    return {
      definitions: {
        ...acc.definitions,
        ...el.definitions,
      },
      paths: {
        ...acc.paths,
        ...el.paths,
      },
    };
  });
  return {
    info: {
      title: options.title,
      version: options.version,
    },
    swagger: '2.0',
    paths,
    definitions,
  };
};

const capitelizeFirst = <Word extends string>(word: Word) =>
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- Capitalize first letter
  (word.slice(0, 1).toUpperCase() + word.slice(1)) as Capitalize<Word>;

const pathToIdentifier = (path: string): string => {
  return capitelizeFirst(
    path
      .replace(/[\/\-](\w?)/g, (_, l: string) => l.toUpperCase())
      .replace(/:(\w?)/g, (_, l: string) => 'By' + l.toUpperCase()),
  );
};

const routeToIdentifier = (route: RouteSubset) => {
  return capitelizeFirst(route.method) + pathToIdentifier(route.path);
};

const getOpenApiPathForRoute = (
  identifier: string,
  route: RouteSubset,
  definitions: OpenAPIV2.DefinitionsObject,
): OpenAPIV2.PathsObject => {
  const operation: OpenAPIV2.OperationObject = {
    operationId: identifier,
    parameters: [
      ...schemaParamToOpenApi(route, 'params', identifier, definitions),
      ...schemaParamToOpenApi(route, 'query', identifier, definitions),
      ...(route.validation.payload
        ? [
            {
              name: 'payload',
              in: 'body',
              schema: {
                $ref: `#/definitions/${schemaToName(identifier, 'payload')}`,
              },
            },
          ]
        : []),
    ],
    responses: {
      default: route.validation.response
        ? {
            description: schemaToName(identifier, 'response'),
            schema: { $ref: `#/definitions/${schemaToName(identifier, 'response')}` },
          }
        : {
            description: 'Unknown response',
          },
    },
  };
  const pathItemObject: OpenAPIV2.PathItemObject = {
    [route.method]: operation,
  };
  const pathsObject: OpenAPIV2.PathsObject = {
    [expressRoutePathToSwaggerPath(route.path)]: pathItemObject,
  };
  return pathsObject;
};

const schemaParamToOpenApi = (
  route: RouteSubset,
  kind: 'query' | 'params',
  identifier: string,
  definitions: OpenAPIV2.DefinitionsObject,
) => {
  const schema = route.validation[kind];
  if (!schema) {
    return [];
  }
  return Object.keys(schema)
    .map((paramName: string): OpenAPIV2.GeneralParameterObject | undefined => {
      const openApiParamIn = kind === 'params' ? 'path' : kind;
      const schema = definitions[schemaToName(identifier, kind)];
      const type = schema?.properties?.[paramName]?.type;
      const enumType = schema?.properties?.[paramName]?.anyOf;

      /* istanbul ignore if */
      if (Array.isArray(type)) {
        // @todo
        return;
      }

      return {
        name: paramName,
        in: openApiParamIn,
        required: schema?.required?.includes(paramName) ?? true,
        type: type ?? 'string',
        ...(enumType && {
          enum: enumType.map((v: IJsonSchema & { readonly const?: string }) => v.const).filter((x): x is string => !!x),
        }),
      };
    })
    .filter((x): x is OpenAPIV2.GeneralParameterObject => typeof x !== 'undefined');
};

const getOpenApiResultForRoute = async (
  identifier: string,
  route: RouteSubset,
): Promise<{ readonly components: { readonly schemas: OpenAPIV2.DefinitionsObject } }> => {
  const data = typesForRoute(route, identifier);

  const reader = getTypeScriptReader({
    nonExported: 'include',
    unsupported: 'warn',
  });
  const writer = getOpenApiWriter({ format: 'json', title: 'My API', version: 'v1', schemaVersion: '2.0' });
  const converter = makeConverter(reader, writer);

  return JSON.parse((await converter.convert({ data })).data);
};

const schemaToName = (identifier: string, kind: keyof RouteSubset['validation']) => {
  const suffix = capitelizeFirst(kind);
  return `${identifier}${suffix}`;
};

const typesForRoute = (route: RouteSubset, identifier: string): string => {
  const typeScriptCode = Object.entries(route.validation).reduce((acc, [kind, maybeSchema]) => {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- it's true
    const name = schemaToName(identifier, kind as keyof RouteSubset['validation']);
    const schema: SomeSchema<any> = isSchema(maybeSchema) ? maybeSchema : object(maybeSchema)();
    return acc + '\n' + `export type ${name} = ${schema.toString()};\n`;
  }, '');
  return Prettier.format(typeScriptCode, { parser: 'typescript' });
};

const expressRoutePathToSwaggerPath = (expressPath: string): string => {
  return expressPath.replace(/:(\w+)/g, '{$1}');
};
