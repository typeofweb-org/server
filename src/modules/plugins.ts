import type { TypeOfWebServerMeta, TypeOfWebRequestMeta } from './augment';
import type { TypeOfWebApp, TypeOfWebServer, HandlerArguments, TypeOfWebCacheConfig } from './shared';
import type { AnyAsyncFunction, MaybeAsync } from '@typeofweb/utils';

type PluginName_ = keyof TypeOfWebServerMeta | keyof TypeOfWebRequestMeta;

export type TypeOfWebServerMetaWithCachedFunctions<PluginName extends keyof TypeOfWebServerMeta> = {
  readonly [K in keyof TypeOfWebServerMeta[PluginName]]: TypeOfWebServerMeta[PluginName][K] extends AnyAsyncFunction
    ?
        | TypeOfWebServerMeta[PluginName][K]
        | { readonly cache: TypeOfWebCacheConfig; readonly fn: TypeOfWebServerMeta[PluginName][K] }
    : TypeOfWebServerMeta[PluginName][K];
};

type PluginCallbackReturnServer<PluginName extends string> = PluginName extends keyof TypeOfWebServerMeta
  ? { readonly server: (server: TypeOfWebServer) => MaybeAsync<TypeOfWebServerMetaWithCachedFunctions<PluginName>> }
  : { readonly server?: never };

type PluginCallbackReturnRequest<PluginName extends string> = PluginName extends keyof TypeOfWebRequestMeta
  ? {
      readonly request: (...args: HandlerArguments) => MaybeAsync<TypeOfWebRequestMeta[PluginName]>;
    }
  : { readonly request?: never };

export type PluginCallbackReturnValue<PluginName extends string> = PluginCallbackReturnServer<PluginName> &
  PluginCallbackReturnRequest<PluginName>;

/**
 * @beta
 */
export interface TypeOfWebPlugin<PluginName extends string> {
  readonly name: PluginName;
  readonly cb: (
    app: TypeOfWebApp,
  ) => PluginName extends PluginName_
    ? MaybeAsync<PluginCallbackReturnValue<PluginName>>
    : MaybeAsync<undefined | void>;
}

export interface TypeOfWebPluginInternal<PluginName extends string> {
  readonly name: PluginName;
  readonly value: void | undefined | PluginCallbackReturnValue<PluginName>;
}

/**
 * @beta
 */
export function createPlugin<PluginName extends string>(
  name: PluginName,
  cb: TypeOfWebPlugin<PluginName>['cb'],
): TypeOfWebPlugin<PluginName> {
  return {
    name,
    cb,
  };
}
