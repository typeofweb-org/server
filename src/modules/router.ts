import { object, validate, ValidationError } from '@typeofweb/schema';
import Express from 'express';

import { HttpError, isStatusError, tryCatch } from '../utils/errors';

import { HttpStatusCode } from './httpStatusCodes';

import type { Json, MaybeAsync } from '../utils/types';
import type { TypeOfWebRequestMeta } from './augment';
import type { HttpMethod } from './httpStatusCodes';
import type { TypeOfWebPluginInternal } from './plugins';
import type { TypeOfWebRequest, TypeOfWebRoute, TypeOfWebServer } from './shared';
import type { SchemaRecord, TypeOfRecord } from './validation';
import type { SomeSchema, TypeOf } from '@typeofweb/schema';

export type ParseRouteParams<Path> = string extends Path
  ? string
  : Path extends `${string}/:${infer Param}/${infer Rest}`
  ? Param | ParseRouteParams<`/${Rest}`>
  : Path extends `${string}:${infer LastParam}`
  ? LastParam
  : never;

export const initRouter = ({
  routes,
  server,
  plugins,
}: {
  readonly routes: readonly TypeOfWebRoute[];
  readonly server: TypeOfWebServer;
  readonly plugins: ReadonlyArray<TypeOfWebPluginInternal<string>>;
}) => {
  const router = Express.Router();

  // @todo sort
  routes.forEach((route) => {
    router[route.method](route.path, finalErrorGuard(routeToExpressHandler({ route, server, plugins })));
  });

  router.use(errorMiddleware(server));
  return router;
};

export const validateRoute = (route: TypeOfWebRoute): boolean => {
  const segments = route.path.split('/');

  const eachRouteSegmentHasAtMostOneParam = segments.every((segment) => (segment.match(/:/g) ?? []).length <= 1);
  if (!eachRouteSegmentHasAtMostOneParam) {
    throw new Error(`RouteValidationError: Each path segment can contain at most one param.`);
  }

  const routeDoesntHaveRegexes = segments.every((segment) => !segment.endsWith(')'));
  if (!routeDoesntHaveRegexes) {
    throw new Error(`RouteValidationError: Don't use regular expressions in routes. Use validators instead.`);
  }

  return true;
};

export const errorMiddleware =
  (server: TypeOfWebServer) =>
  (err: unknown, _req: Express.Request, res: Express.Response, next: Express.NextFunction) => {
    server.events.emit(':error', err);
    if (res.headersSent) {
      next(err);
      return;
    }

    if (!err) {
      return res.status(0).end();
    }

    if (err instanceof ValidationError) {
      return res.status(400).json({ name: err.name, message: err.message, body: err.details });
    }

    if (err instanceof HttpError) {
      return res.status(err.statusCode).json({ name: err.name, message: err.message, body: err.body });
    }

    if (isStatusError(err)) {
      return res.status(err.statusCode).json({ name: HttpStatusCode[err.statusCode], body: err });
    }

    // @todo if (DEBUG)
    return res.status(500).json(err);
  };

type AsyncHandler = (
  ...args: Parameters<Express.Handler>
) => Promise<ReturnType<Express.Handler>> | ReturnType<Express.Handler>;

const finalErrorGuard = (h: AsyncHandler): AsyncHandler => {
  return async (req, res, next) => {
    try {
      await h(req, res, next);
    } catch (err) {
      next(err);
    }
  };
};

export const routeToExpressHandler = <
  Path extends string,
  ParamsKeys extends ParseRouteParams<Path>,
  Params extends SchemaRecord<ParamsKeys>,
  Query extends SchemaRecord<string>,
  Payload extends SomeSchema<Json>,
  Response extends SomeSchema<Json>,
>({
  plugins,
  route,
  server,
}: {
  readonly plugins: ReadonlyArray<TypeOfWebPluginInternal<string>>;
  readonly route: {
    readonly path: Path;
    readonly method: HttpMethod;
    readonly validation: {
      readonly params?: Params;
      readonly query?: Query;
      readonly payload?: Payload;
      readonly response?: Response;
    };
    handler(
      request: TypeOfWebRequest<Path, TypeOfRecord<Params>, TypeOfRecord<Query>, TypeOf<Payload>>,
    ): MaybeAsync<TypeOf<Response>>;
  };
  readonly server: TypeOfWebServer;
}): AsyncHandler => {
  return async (req, res, next) => {
    const params = tryCatch(() =>
      route.validation.params ? validate(object(route.validation.params)())(req.params) : req.params,
    );
    if (params._t === 'left') {
      return next(new HttpError(HttpStatusCode.BadRequest, HttpStatusCode[HttpStatusCode.BadRequest], params.value));
    }

    const query = tryCatch(() =>
      route.validation.query ? validate(object(route.validation.query)())(req.query) : req.query,
    );
    if (query._t === 'left') {
      return next(new HttpError(HttpStatusCode.BadRequest, HttpStatusCode[HttpStatusCode.BadRequest], query.value));
    }

    const payload = tryCatch(() =>
      route.validation.payload ? validate(route.validation.payload)(req.body) : req.body,
    );
    if (payload._t === 'left') {
      return next(new HttpError(HttpStatusCode.BadRequest, HttpStatusCode[HttpStatusCode.BadRequest], payload.value));
    }

    const request: TypeOfWebRequest<Path, TypeOfRecord<Params>, TypeOfRecord<Query>, TypeOf<Payload>> = {
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- these types are validated
      params: params.value as TypeOfRecord<Params>,
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- these types are validated
      query: query.value as TypeOfRecord<Query>,
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- these types are validated
      payload: payload.value as TypeOf<Payload>,

      server,
      plugins: [],
      path: route.path,
      _rawReq: req,
      _rawRes: res,
    };

    await plugins.reduce(async (acc, plugin) => {
      if (typeof plugin?.value?.request !== 'function') {
        return acc;
      }

      await acc;

      // @ts-expect-error
      const requestMetadata = await plugin.value.request(request);
      if (requestMetadata) {
        // @ts-expect-error
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- ok
        request.plugins[plugin.name as keyof TypeOfWebRequestMeta] = requestMetadata;
      }
    }, Promise.resolve());

    server.events.emit(':request', request);

    const originalResult = await route.handler(request);

    const result = tryCatch(() =>
      route.validation.response ? validate(route.validation.response)(originalResult) : originalResult,
    );

    if (result._t === 'left') {
      const err = result.value;
      if (err instanceof ValidationError) {
        // @todo don't send stacktrace on production
        return next(new HttpError(HttpStatusCode.InternalServerError, err.message, err.details));
      }
      return next(
        new HttpError(
          HttpStatusCode.InternalServerError,
          HttpStatusCode[HttpStatusCode.InternalServerError],
          result.value,
        ),
      );
    }

    if (res.headersSent) {
      return;
    }

    server.events.emit(':response', result.value);

    if (result.value === null) {
      res.status(204).end();
      return;
    } else if (result.value === undefined) {
      console.warn(
        'Handler returned `undefined` which usually means you forgot to `await` something. If you want an empty response, return `null` instead.',
      );
      res.status(204).end();
      return;
    } else {
      res.status(200).json(result.value);
      return;
    }
  };
};
