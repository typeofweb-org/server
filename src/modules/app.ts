import { URL } from 'url';

import CacheManager from 'cache-manager';
import CookieParser from 'cookie-parser';
import Cors from 'cors';
import Supertest from 'supertest';

import { deepMerge } from '../utils/merge';
import { promiseOriginFnToNodeCallback } from '../utils/node';
import { generateServerId } from '../utils/uniqueId';

import { createCachedFunction } from './cache';
import { createEventBus } from './events';
import { initApp, listenExpressServer } from './http';
import { getOpenApiForRoutes } from './openapi';
import { initRouter, validateRoute } from './router';

import type { AnyFunction, DeepPartial, DeepWritable, JsonPrimitive, MaybeAsync } from '../utils/types';
import type { TypeOfWebServerMeta } from './augment';
import type { TypeOfWebPluginInternal } from './plugins';
import type { AppOptions, TypeOfWebRoute, TypeOfWebApp, TypeOfWebServer, TypeOfWebCacheConfig } from './shared';

const defaultAppOptions: AppOptions = {
  hostname: 'localhost',
  port: 3000,
  cors: {
    origin: true,
    credentials: true,
  },
  cookies: {
    encrypted: true,
    secure: true,
    httpOnly: true,
    sameSite: 'lax',
    secret: '',
  },
  router: {
    strictTrailingSlash: false,
  },
  openapi: false,
};

export function createApp(opts: DeepPartial<AppOptions>): TypeOfWebApp {
  const options = deepMerge(opts, defaultAppOptions);
  const memoryCache = CacheManager.caching({ store: 'memory', ttl: 0 });

  const server: DeepWritable<TypeOfWebServer> = {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- these properties are supposed to be added by the plugins inside `async start()`
    plugins: {} as TypeOfWebServer['plugins'],
    events: createEventBus(),
    /* istanbul ignore next */
    get address() {
      /* istanbul ignore next */
      return null;
    },

    id: generateServerId(),
  };

  /* eslint-disable functional/prefer-readonly-type -- ok */
  const routes: Array<TypeOfWebRoute> = [];
  /* eslint-disable functional/prefer-readonly-type -- ok */
  const plugins: Array<TypeOfWebPluginInternal<string>> = [];

  let mutableIsInitialized = false;

  function initServerPlugins() {
    return plugins.reduce(async (acc, plugin) => {
      if (!plugin?.value || typeof plugin?.value.server !== 'function') {
        return acc;
      }

      await acc;

      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- ok
      const pluginServer = plugin.value.server as unknown as (server: TypeOfWebServer) => MaybeAsync<
        Record<
          string,
          | JsonPrimitive
          | AnyFunction
          | {
              readonly cache: TypeOfWebCacheConfig;
              readonly fn: AnyFunction;
            }
        >
      >;

      const result = await pluginServer(server);
      const serverMetadata = !result
        ? null
        : // skip iterating over instances of custom classes
        result.constructor !== Object
        ? result
        : Object.fromEntries(
            Object.entries(result).map(([key, val]) => {
              if (typeof val === 'object' && val && 'cache' in val) {
                return [key, createCachedFunction({ ...val, cacheInstance: memoryCache })];
              }
              return [key, val];
            }),
          );

      if (serverMetadata) {
        // @ts-expect-error
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- if serverMetadata exists then plugin name is keyof TypeOfWebServerMeta
        server.plugins[plugin.name as keyof TypeOfWebServerMeta] = serverMetadata;
      }
    }, Promise.resolve());
  }

  async function initialize() {
    if (mutableIsInitialized) {
      return;
    }

    if (options.cors) {
      const origin =
        typeof options.cors.origin === 'function'
          ? promiseOriginFnToNodeCallback(options.cors.origin)
          : options.cors.origin;
      app._rawExpressApp.use(
        Cors({
          origin,
          credentials: options.cors.credentials,
        }),
      );
    }

    app._rawExpressApp.use(CookieParser(''));

    await initServerPlugins();

    app._rawExpressRouter = initRouter({ server, appOptions: options, routes, plugins });
    app._rawExpressApp.use(app._rawExpressRouter);

    /* istanbul ignore if */
    if (options.openapi) {
      const SwaggerUiExpress = await import('swagger-ui-express');
      const openapi = await getOpenApiForRoutes(routes, options.openapi);
      const path = options.openapi.path ?? '/documentation';

      app._rawExpressRouter.use(path, SwaggerUiExpress.serve);
      app._rawExpressRouter.get(path, SwaggerUiExpress.setup(openapi));
    }

    mutableIsInitialized = true;
    server.events.emit(':server', server);
  }

  const app: DeepWritable<TypeOfWebApp> = {
    _rawExpressApp: initApp(),
    events: server.events,

    async plugin(plugin) {
      const pluginDefinition = {
        name: plugin.name,
        value: await plugin.cb(app),
      };
      plugins.push(pluginDefinition);
      return app;
    },

    route(route) {
      validateRoute(route);
      routes.push(route);
      return app;
    },

    async inject(injection) {
      await initialize();

      let mutableTest = Supertest(app._rawExpressApp)[injection.method](injection.path);

      if (injection.payload) {
        mutableTest = mutableTest.send(injection.payload);
      }

      if (injection.headers) {
        mutableTest = Object.entries(injection.headers).reduce(
          (acc, [header, value]) => acc.set(header, value),
          mutableTest,
        );
      }

      if (injection.cookies) {
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- string[]
        mutableTest = mutableTest.set('Cookie', injection.cookies as string[]);
      }

      const result = await mutableTest;
      return result;
    },

    /* istanbul ignore next */
    async start() {
      await initialize();
      app._rawExpressServer = await listenExpressServer(app._rawExpressApp, options);

      Object.defineProperty(server, 'address', {
        get() {
          const address = app._rawExpressServer?.address();
          if (typeof address !== 'object' || !address) {
            return null;
          }

          const host = address.family === 'IPv6' ? `[${address.address}]` : address.address;

          return new URL(`http://${host}:${address.port}`);
        },
      });

      return server;
    },

    /* istanbul ignore next */
    stop() {
      return new Promise((resolve, reject) => {
        if (!app._rawExpressServer) {
          return resolve();
        }
        return app._rawExpressServer.stop((err) => {
          app._rawExpressServer = undefined;
          if (err) {
            return reject(err);
          }
          return resolve();
        });
      });
    },
  };

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- readonly
  return app as TypeOfWebApp;
}
