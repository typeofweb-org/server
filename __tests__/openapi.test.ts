import { array, boolean, number, object, oneOf, string } from '@typeofweb/schema';

import { routeConfigToOpenApiPathsDefinitions } from '../src/modules/openapi';

describe('routeConfigToOpenApi', () => {
  it('should work', async () => {
    await routeConfigToOpenApiPathsDefinitions({
      path: '/users/:userId/invoices',
      method: 'post',
      validation: {
        params: {
          userId: string(),
        },
        payload: object({
          id: string(),
          item: object({
            price: number(),
          })(),
        })(),
        query: {
          isFun: boolean(),
          search: string(),
          category: oneOf(['html', 'css'])(),
        },
        response: array(
          object({
            id: string(),
            total: number(),
          })(),
        )(),
      },
    });
  });
});
