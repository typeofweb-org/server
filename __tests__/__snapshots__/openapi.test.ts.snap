// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`open api generator getOpenApiForRoutes should generate schema for whole app 1`] = `
Object {
  "definitions": Object {
    "GetUsersByUserIdInvoicesParams": Object {
      "additionalProperties": false,
      "properties": Object {
        "userId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "userId",
      ],
      "type": "object",
    },
    "GetUsersByUserIdInvoicesResponse": Object {
      "items": Object {
        "additionalProperties": false,
        "properties": Object {
          "id": Object {
            "type": "string",
          },
          "total": Object {
            "type": "number",
          },
        },
        "required": Array [
          "id",
          "total",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "PostUsersByUserIdInvoicesParams": Object {
      "additionalProperties": false,
      "properties": Object {
        "userId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "userId",
      ],
      "type": "object",
    },
    "PostUsersByUserIdInvoicesPayload": Object {
      "additionalProperties": false,
      "properties": Object {
        "id": Object {
          "type": "string",
        },
        "item": Object {
          "additionalProperties": false,
          "properties": Object {
            "price": Object {
              "type": "number",
            },
          },
          "required": Array [
            "price",
          ],
          "type": "object",
        },
      },
      "required": Array [
        "id",
        "item",
      ],
      "type": "object",
    },
    "PostUsersByUserIdInvoicesQuery": Object {
      "additionalProperties": false,
      "properties": Object {
        "category": Object {
          "anyOf": Array [
            Object {
              "const": "html",
              "type": "string",
            },
            Object {
              "const": "css",
              "type": "string",
            },
          ],
        },
        "isFun": Object {
          "type": "boolean",
        },
        "search": Object {
          "type": "string",
        },
      },
      "required": Array [
        "isFun",
        "search",
        "category",
      ],
      "type": "object",
    },
    "PostUsersByUserIdInvoicesResponse": Object {
      "items": Object {
        "additionalProperties": false,
        "properties": Object {
          "id": Object {
            "type": "string",
          },
          "total": Object {
            "type": "number",
          },
        },
        "required": Array [
          "id",
          "total",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "PutUsersByUserIdInvoicesParams": Object {
      "additionalProperties": false,
      "properties": Object {
        "userId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "userId",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "title": "Swagger",
    "version": "1.1.1",
  },
  "paths": Object {
    "/users/{userId}/invoices": Object {
      "put": Object {
        "operationId": "PutUsersByUserIdInvoices",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "required": true,
            "type": "string",
          },
        ],
        "responses": Object {
          "default": Object {
            "description": "Unknown response",
          },
        },
      },
    },
  },
  "swagger": "2.0",
}
`;

exports[`open api generator routeConfigToOpenApiPathsDefinitions should generate single schema 1`] = `
Object {
  "definitions": Object {
    "PostUsersByUserIdInvoicesParams": Object {
      "additionalProperties": false,
      "properties": Object {
        "userId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "userId",
      ],
      "type": "object",
    },
    "PostUsersByUserIdInvoicesPayload": Object {
      "additionalProperties": false,
      "properties": Object {
        "id": Object {
          "type": "string",
        },
        "item": Object {
          "additionalProperties": false,
          "properties": Object {
            "price": Object {
              "type": "number",
            },
          },
          "required": Array [
            "price",
          ],
          "type": "object",
        },
      },
      "required": Array [
        "id",
        "item",
      ],
      "type": "object",
    },
    "PostUsersByUserIdInvoicesQuery": Object {
      "additionalProperties": false,
      "properties": Object {
        "category": Object {
          "anyOf": Array [
            Object {
              "const": "html",
              "type": "string",
            },
            Object {
              "const": "css",
              "type": "string",
            },
          ],
        },
        "isFun": Object {
          "type": "boolean",
        },
        "search": Object {
          "type": "string",
        },
      },
      "required": Array [
        "isFun",
        "search",
        "category",
      ],
      "type": "object",
    },
    "PostUsersByUserIdInvoicesResponse": Object {
      "items": Object {
        "additionalProperties": false,
        "properties": Object {
          "id": Object {
            "type": "string",
          },
          "total": Object {
            "type": "number",
          },
        },
        "required": Array [
          "id",
          "total",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "paths": Object {
    "/users/{userId}/invoices": Object {
      "post": Object {
        "operationId": "PostUsersByUserIdInvoices",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "required": true,
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "isFun",
            "required": true,
            "type": "boolean",
          },
          Object {
            "in": "query",
            "name": "search",
            "required": true,
            "type": "string",
          },
          Object {
            "enum": Array [
              "html",
              "css",
            ],
            "in": "query",
            "name": "category",
            "required": true,
            "type": "string",
          },
          Object {
            "in": "body",
            "name": "payload",
            "schema": Object {
              "$ref": "#/definitions/PostUsersByUserIdInvoicesPayload",
            },
          },
        ],
        "responses": Object {
          "default": Object {
            "description": "PostUsersByUserIdInvoicesResponse",
            "schema": Object {
              "$ref": "#/definitions/PostUsersByUserIdInvoicesResponse",
            },
          },
        },
      },
    },
  },
}
`;
